### PR TITLE
Implement term handling & docs

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

* Add SparseTerm
* Add PauliTermVec, IndexVec, BitTermVec
* Improve tests
* Add doxygen docs consistently

### CODE REVIEW

File: crates/accelerate/src/py_sparse_observable.rs [hunk 1]
```diff
@@ -395,6 +395,402 @@ fn make_py_bit_term(py: Python) -> PyResult<Py<PyType>> {
     Ok(obj.downcast_into::<PyType>()?.unbind())
 }
 
+/// An observable over Pauli bases that stores its data in a qubit-sparse format.
```
Comment (mrossinek) [inline crates/accelerate/src/py_sparse_observable.rs:398]: Should we remove this docstring from the pure Rust version? Or do you want it duplicated entirely?

File: crates/c_ext/src/sparse_observable.rs [hunk 2]
```diff
@@ -11,13 +11,290 @@
...
+    values: Vec<BitTerm>,
+}
+
+/// A Pauli term vector, containing ``(index, bit_term)`` tuples.
+#[derive(Debug, Clone)]
+pub struct PauliTermVec {
+    // We store the data as two separate vectors, which simplified constructing
+    // SparseTerms later on, which expect this format.
+    indices: Vec<u32>,
+    bit_terms: Vec<BitTerm>,
```
Comment (mrossinek) [inline crates/c_ext/src/sparse_observable.rs:22]:
```suggestion
    indices: IndexVec,
    bit_terms: BitTermVec,
```
Or do you see a reason not to do this?

File: crates/c_ext/src/sparse_observable.rs [hunk 3]
```diff
@@ -11,13 +11,290 @@
...
+///  
+///     SparseTerm *term = obs_term(obs, 0);  // get the 0th Pauli term in the observable
+///     PauliTerm *pauli = obsterm_pauli(term, 0);  // get the 0th Pauli in the Pauli term
+///     printf("Bit term: %i, Index: %i", pauli->bit_term, pauli->index);
+///
+///     pauli_deallocate(pauli);  // deallocate the struct
+///
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub extern "C" fn pauli_deallocate(pauli: &mut PauliTerm) {
```
Comment (mrossinek) [inline crates/c_ext/src/sparse_observable.rs]:
```suggestion
pub extern "C" fn pauli_free(pauli: &mut PauliTerm) {
```
I recall you mentioned wanting to rename these functions like so, no?

File: test/c/test_sparse_observable.c [hunk 4]
```diff
@@ -148,22 +179,127 @@ int test_num_qubits()
...
 {
     int num_failed = 0;
-
     num_failed += RUN_TEST(test_zero);
     num_failed += RUN_TEST(test_identity);
     num_failed += RUN_TEST(test_add);
-    num_failed += RUN_TEST(test_mult_real);
-    num_failed += RUN_TEST(test_mult_complex);
+    num_failed += RUN_TEST(test_mult);
+    fflush(stderr);
```
Comment (mrossinek) [inline test/c/test_sparse_observable.c:276]: I think you accidentally moved this line up
```suggestion
```

File: test/c/test_sparse_observable.c [hunk 5]
```diff
@@ -148,22 +179,127 @@ int test_num_qubits()
...
     num_failed += RUN_TEST(test_canonicalize);
     num_failed += RUN_TEST(test_copy);
     num_failed += RUN_TEST(test_num_terms);
     num_failed += RUN_TEST(test_num_qubits);
+    num_failed += RUN_TEST(test_custom_build);
+    num_failed += RUN_TEST(test_term);
 
     fprintf(stderr, "=== Number of failed subtests: %i\n", num_failed);
```
Comment (mrossinek) [inline test/c/test_sparse_observable.c:284]:
```suggestion
    fprintf(stderr, "=== Number of failed subtests: %i\n", num_failed);
    fflush(stderr);
```

**Review by mrossinek — APPROVED** [review]
A more thorough docs review can follow later

File: crates/accelerate/src/py_sparse_observable.rs (hunk 1 above)
Comment (Cryoris) [reply crates/accelerate/src/py_sparse_observable.rs:398]: We should strip it and adjust it to Rust. It currently contains a lot of Python-specific statements.

File: crates/c_ext/src/sparse_observable.rs (hunk 2 above)
Comment (Cryoris) [reply crates/c_ext/src/sparse_observable.rs:22]: `IndexVec & BitTermVec` are removed now

File: test/c/test_sparse_observable.c (hunk 4 above)
Comment (Cryoris) [reply test/c/test_sparse_observable.c:276]: yeah I was trying to get C's `printf` to print interleaved with calls to Rust's `println` but that didn't work... I'll revert 🙂

**Pull request merged**